pkg/scheduler/queue/queue_test.go

```diff
@@ -63,10 +78,149 @@ func makeSchedulerRequest(tenantID string) *SchedulerRequest {
+// The queue broker then round-robins between the split queues, which has the effect of alternating between
+// dequeuing the slow queries and normal queries rather than blocking normal queries behind slow queries.
+func TestMultiDimensionalQueueFairnessSlowConsumerEffects(t *testing.T) {
+	promRegistry := prometheus.NewPedanticRegistry()
+
+	maxQueriersPerTenant := 0 // disable shuffle sharding
+	forgetQuerierDelay := time.Duration(0)
+	maxOutstandingRequestsPerTenant := 1000
+
+	totalRequests := 100
+	numTenants := 1
+	numProducers := 10
+	numConsumers := 1
+
+	normalQueueDimension := storeGatewayQueueDimension
+	slowConsumerLatency := 20 * time.Millisecond
+	slowConsumerQueueDimension := ingesterQueueDimension
+	normalQueueDimensionFunc := func() []string { return []string{normalQueueDimension} }
+	slowQueueDimensionFunc := func() []string { return []string{slowConsumerQueueDimension} }
+
+	additionalQueueDimensionsEnabledCases := []bool{false, true}
+	queueDurationTotals := map[bool]map[string]float64{
+		false: {normalQueueDimension: 0.0, slowConsumerQueueDimension: 0.0},
+		true:  {normalQueueDimension: 0.0, slowConsumerQueueDimension: 0.0},
+	}
+
+	for _, additionalQueueDimensionsEnabled := range additionalQueueDimensionsEnabledCases {
+
+		// production code uses a histogram for queue duration, but the counter is a more direct metric for test cases
+		queueDuration := promauto.With(promRegistry).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_query_scheduler_queue_duration_total_seconds",
+			Help: "Total time spent by requests in queue before getting picked up by a querier.",
+		}, []string{"additional_queue_dimensions"})
+
+		queue := NewRequestQueue(
+			log.NewNopLogger(),
+			maxOutstandingRequestsPerTenant,
+			additionalQueueDimensionsEnabled,
+			forgetQuerierDelay,
+			promauto.With(nil).NewGaugeVec(prometheus.GaugeOpts{}, []string{"tenant"}),
+			promauto.With(nil).NewCounterVec(prometheus.CounterOpts{}, []string{"tenant"}),
+			promauto.With(nil).NewHistogram(prometheus.HistogramOpts{}),
+		)
+
+		ctx := context.Background()
+		require.NoError(t, queue.starting(ctx))
+		t.Cleanup(func() {
+			require.NoError(t, queue.stop(nil))
+		})
+
+		// fill queue first with the slow queries, then the normal queries
+		for _, queueDimensionFunc := range []func() []string{slowQueueDimensionFunc, normalQueueDimensionFunc} {
+			startProducersChan := make(chan struct{})
+			producersErrGroup, _ := errgroup.WithContext(ctx)
+
+			runProducer := runQueueProducerIters(
+				queue, maxQueriersPerTenant, totalRequests/2, numProducers, numTenants, startProducersChan, queueDimensionFunc,
+			)
+			for producerIdx := 0; producerIdx < numProducers; producerIdx++ {
+				producerIdx := producerIdx
+				producersErrGroup.Go(func() error {
+					return runProducer(producerIdx)
+				})
+			}
+			close(startProducersChan)
+			err := producersErrGroup.Wait()
+			if err != nil {
+				require.NoError(t, err)
+			}
```

Comment:
Given we want to fill the queue ahead of time, and don't need to do anything like waiting for some requests to be consumed, could we just directly add requests to the queue here with `queueProduce`? This would remove the need for the bookkeeping introduced by the channels and `WaitGroup` etc.

Reply:
I'll check - I originally did this because I was adding >1000 requests and the time it took to add them all to the queue with just a loop was long enough that it skewed the timing results at at the end.